app/services/fact_importer.rb

```diff
@@ -39,31 +39,86 @@ def initialize(host, facts = {})
 
   # expect a facts hash
   def import!
+    ensure_fact_names
+
     ActiveRecord::Base.transaction do
       delete_removed_facts
       update_facts
       add_new_facts
     end
 
-    if @error
-      Foreman::Logging.exception("Error during fact import for #{@host.name}", @error)
-      raise ::Foreman::WrappedException.new(@error, N_("Import of facts failed for host %s"), @host.name)
-    end
+    report_error(@error) if @error
     logger.info("Import facts for '#{host}' completed. Added: #{counters[:added]}, Updated: #{counters[:updated]}, Deleted #{counters[:deleted]} facts")
   end
 
+  def report_error(error)
+    Foreman::Logging.exception("Error during fact import for #{@host.name}", error)
+    raise ::Foreman::WrappedException.new(error, N_("Import of facts failed for host %s"), @host.name)
+  end
+
   # to be defined in children
   def fact_name_class
     raise NotImplementedError
   end
 
   private
 
-  attr_reader :host, :facts
+  attr_reader :host, :facts, :fact_names, :fact_names_by_id
+
+  def ensure_fact_names
+    initialize_fact_names
+
+    missing_keys = facts.keys - fact_names.keys
+    add_missing_fact_names(missing_keys)
+  end
+
+  def add_missing_fact_names(missing_keys)
+    missing_keys.sort.each do |fact_name|
+      # create a new record and make sure it could be saved.
+      name_record = fact_name_class.new(fact_name_attributes(fact_name))
+
+      ActiveRecord::Base.transaction(:requires_new => true) do
+        begin
+          unless name_record.save
+            throw ActiveRecord::RecordInvalid.new(name_record) unless duplicate_name_error?(name_record)
+
+            # if the record could not be saved, then it should already be in the database
+            # throw an exception if it isn't
+            name_record = fact_name_class.find_by!(name: fact_name)
+          end
+        rescue ActiveRecord::RecordNotUnique
+          name_record = fact_name_class.find_by!(name: fact_name)
+        end
+      end
+
+      # make the new record available immediately for other fact_name_attributes calls
+      @fact_names[fact_name] = name_record
+    end
+  end
+
+  def duplicate_name_error?(name_record)
+    name_record.errors.size == 1 && name_record.errors.details[:name].first.try(:[], :error) == :taken
+  end
```

Comment:
> Current SQL standards support atomic operation to create or find a record in database

Unfortunately it's not SQL standard, and will lock us to PG (especially a quite advanced version), which IMHO should be avoided (even at a cost of couple ms at each request).

The main problem here is not the statement that will fail (even without a transaction), but the connection that will be rendered useless, since until the connection is released it will be in a failed state. One of the errors you were seeing was indeed about trying to save a record from such connection. This way I keep the damage contained to the single statement.

As a general statement, I really want to get the whole facts processing out of foreman into a separate process - it will be a lot slimmer than spawning a new full blown rails app for each fact report. Maybe even consider a different storage type for this kind of information. Although it's a long term goal, and I don't think this PR is the place to start such a discussion.

Reply:
You can now delete `duplicate_name_error` which is not used anymore.